crates/languages/src/python.rs

```diff
@@ -678,28 +704,68 @@ impl ToolchainLister for PythonToolchainProvider {
-                                    (true, false) => Ordering::Less,
-                                    (false, true) => Ordering::Greater,
-                                    _ => Ordering::Equal,
-                                }
-                            })
-                            .unwrap_or(Ordering::Equal)
-                    } else {
-                        Ordering::Equal
-                    }
-                })
+            // Compare venv names against worktree .venv file
+            let venv_ordering =
+                wr_venv
+                    .as_ref()
+                    .map_or(Ordering::Equal, |venv| match (&lhs.name, &rhs.name) {
+                        (Some(l), Some(r)) => match (l == venv, r == venv) {
+                            (true, false) => Ordering::Less,
+                            (false, true) => Ordering::Greater,
+                            _ => Ordering::Equal,
+                        },
+                        (Some(l), None) if l == venv => Ordering::Less,
+                        (None, Some(r)) if r == venv => Ordering::Greater,
+                        _ => Ordering::Equal,
+                    });
+
+            // Compare project paths against worktree root
+            let proj_ordering = match (&lhs.project, &rhs.project) {
+                (Some(l), Some(r)) => match (l == &wr, r == &wr) {
+                    (true, false) => Ordering::Less,
+                    (false, true) => Ordering::Greater,
+                    _ => Ordering::Equal,
+                },
+                (Some(l), None) if l == &wr => Ordering::Less,
+                (None, Some(r)) if r == &wr => Ordering::Greater,
+                _ => Ordering::Equal,
+            };
+
+            // Compare environment priorities
+            let priority_ordering = env_priority(lhs.kind).cmp(&env_priority(rhs.kind));
+
+            // Compare conda prefixes
+            let conda_ordering = if lhs.kind == Some(PythonEnvironmentKind::Conda) {
+                environment
+                    .get_env_var("CONDA_PREFIX".to_string())
+                    .map(|conda_prefix| {
+                        let is_match = |exe: &Option<PathBuf>| {
+                            exe.as_ref().map_or(false, |e| e.starts_with(&conda_prefix))
+                        };
+                        match (is_match(&lhs.executable), is_match(&rhs.executable)) {
+                            (true, false) => Ordering::Less,
+                            (false, true) => Ordering::Greater,
+                            _ => Ordering::Equal,
+                        }
+                    })
+                    .unwrap_or(Ordering::Equal)
+            } else {
+                Ordering::Equal
+            };
+
+            venv_ordering
+                .then(proj_ordering)
+                .then(priority_ordering)
+                .then(conda_ordering)
```

Comment:
Could we use `then_with` for these instead? That way we won't have to execute all comparisons if any of the subsequent ones bails early.

Reply:
Done.